### ISSUE TITLE
Different results for NUTS on different systems

### ISSUE THREAD

**Issue by hriebl** [blackjax-devs/blackjax#181]
### Bug Description

Dear Rémi,

a colleague and me noticed that the BlackJAX NUTS kernel returns numerically different results on our systems. The differences might seem tiny, but we found that they can add up over a long chain. He's on macOS, and I'm using Fedora. For the sake of reproducibility, shouldn't the results be exactly the same? Do you have any ideas what might cause the differences?

Best,
Hannes

#### Steps/Code to Reproduce

``` python
import blackjax.nuts as nuts
import jax
import jax.numpy as jnp
import numpy as np

n = 30
rng = np.random.default_rng(1337)
x = rng.normal(0.0, 1.0, size=n)


def log_prob(loc):
    return jnp.sum(jax.scipy.stats.norm.logpdf(x, loc, 1.0))


state = nuts.new_state(0.0, log_prob)
kernel = nuts.kernel(log_prob, step_size=0.2, inverse_mass_matrix=jnp.array([1.0]))
kernel = jax.jit(kernel)

m = 10
key = jax.random.PRNGKey(1337)
key = jax.random.split(key, m)

if __name__ == "__main__":
    for i in range(m):
        state, info = kernel(key[i], state)
        print(f"[iter {i}] position = {state.position.item()}")
        print(f"[iter {i}] acc prob = {info.acceptance_probability.item()}")
```

#### Expected Results

Exactly the same results on both systems.

#### Actual Results

Differences: <https://www.diffchecker.com/dj3ZKduJ>

#### Versions

His system: MacBook Pro (Retina, 13 inch, late 2013), macOS 11.6.4, Python 3.10.1, JAX 0.3.1, jaxlib 0.3.0, BlackJAX 0.3.0
My system: Dell Latitude 5420, Fedora 35, Python 3.10.2, JAX 0.3.1, jaxlib 0.3.0, BlackJAX 0.3.0

**Comment by sethaxen** [comment]
I seem to remember a Twitter thread by Michael Betancourt where he argued out that in general one can only guarantee reproducibility with HMC if the same machine, compiler, and OS is used, that minute floating point differences across different machines ultimately can add up to major differences in the transitions, i.e. I guess you're targeting somewhat different shadow Hamiltonians.

**Comment by wiep** [comment]
Using Python 3.10.2 on the MacBook Pro, I see the same results as before. I also tried running the snippet on another more recent MacBook Pro. Same results.

However, when I run the snipped on colab, the results are different from all the others. Also, CPU and
 GPU differ on colab.

Using double precision, I can still observe differences although they are smaller (see the screenshots below);

It might be just a coincidence that two MacBooks produce the same results, and usually, one can not rely on reproducibility among different machines.

![colab_gpu_cpu_64](https://user-images.githubusercontent.com/603509/155390943-44562d6f-e31c-469b-8bf2-d17733a2d0d2.png)
![mac_colab_64](https://user-images.githubusercontent.com/603509/155390949-cd0c5cde-b0a2-45b5-afc7-5d38ff1c86b4.png)

**Comment by hriebl** [comment]
Thank you, @sethaxen, for the hint! I just found [the section on reproducibility in the Stan manual](https://mc-stan.org/docs/2_29/reference-manual/reproducibility.html). Indeed, they are saying that reproducibility can only be guaranteed if all of these are the same:

- Stan version
- Stan interface (RStan, PyStan, CmdStan) and version, plus version of interface language (R, Python, shell)
- versions of included libraries (Boost and Eigen)
- operating system version
- computer hardware including CPU, motherboard and memory
- C++ compiler, including version, compiler flags, and linked libraries
- same configuration of call to Stan, including random seed, chain ID, initialization and data

Pretty depressing, huh? And as Stan and XLA both depend on Eigen, I guess similar considerations apply to BlackJAX and JAX in general...

**Comment by rlouf** [comment]
Yes that's a shame, but you can also see it as the ultimate test for the quality of your samples: if your colleague can't reproduce the results you are probably doing it wrong :)

Closing as it seems there is nothing we can do about it.
